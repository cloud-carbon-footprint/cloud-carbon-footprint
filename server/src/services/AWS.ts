import { CloudWatch, CostExplorer } from 'aws-sdk'
import { AWS_REGIONS } from './AWSRegions'

export async function getCostAndUsageResponses(
  params: CostExplorer.GetCostAndUsageRequest,
): Promise<CostExplorer.GetCostAndUsageResponse[]> {
  const costExplorer = new CostExplorer({
    region: AWS_REGIONS.US_EAST_1, //must be us-east-1 to work
  })

  let response: CostExplorer.GetCostAndUsageResponse = {}
  const responses: CostExplorer.GetCostAndUsageResponse[] = []

  do {
    response = await costExplorer.getCostAndUsage({ ...params, NextPageToken: response.NextPageToken }).promise()
    responses.push(response)
  } while (response.NextPageToken)
  return responses
}

export async function getMetricDataResponses(
  params: CloudWatch.GetMetricDataInput,
): Promise<CloudWatch.GetMetricDataOutput[]> {
  const cloudWatch = new CloudWatch()

  let response: CloudWatch.GetMetricDataOutput = {}
  const responses: CloudWatch.GetMetricDataOutput[] = []

  do {
    response = await cloudWatch.getMetricData({ ...params, NextToken: response.NextToken }).promise()
    responses.push(response)
  } while (response.NextToken)

  return responses
}
