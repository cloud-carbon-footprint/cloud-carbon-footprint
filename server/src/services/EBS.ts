import ICloudService from '@domain/ICloudService'
import {
  DiskType,
  getEstimatesFromCostExplorer,
  getUsageFromCostExplorer,
  VolumeUsage,
} from '@services/StorageUsageMapper'
import FootprintEstimate from '@domain/FootprintEstimate'

export default class EBS implements ICloudService {
  serviceName = 'ebs'

  async getEstimates(start: Date, end: Date, region: string): Promise<FootprintEstimate[]> {
    const usage: VolumeUsage[] = await this.getUsage(start, end, region)
    return getEstimatesFromCostExplorer(start, end, region, usage)
  }

  async getUsage(startDate: Date, endDate: Date, region: string): Promise<VolumeUsage[]> {
    const params = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: [
                'EC2: EBS - SSD(gp2)',
                'EC2: EBS - SSD(io1)',
                'EC2: EBS - HDD(sc1)',
                'EC2: EBS - HDD(st1)',
                'EC2: EBS - Magnetic',
              ],
            },
          },
          { Dimensions: { Key: 'REGION', Values: [region] } },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['UsageQuantity'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return await getUsageFromCostExplorer(params, this.getDiskType, region)
  }

  private getDiskType = (awsGroupKey: string) => {
    if (awsGroupKey.endsWith('VolumeUsage.gp2') || awsGroupKey.endsWith('VolumeUsage.piops')) return DiskType.SSD
    if (
      awsGroupKey.endsWith('VolumeUsage.st1') ||
      awsGroupKey.endsWith('VolumeUsage.sc1') ||
      awsGroupKey.endsWith('VolumeUsage')
    )
      return DiskType.HDD
    console.warn('Unexpected Cost explorer Dimension Name: ' + awsGroupKey)
  }
}
