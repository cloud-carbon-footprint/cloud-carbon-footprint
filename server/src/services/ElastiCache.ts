import { config } from 'aws-sdk'
import ComputeUsage from '@domain/ComputeUsage'
import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import { getComputeUsage } from '@services/ComputeUsageMapper'
import { CACHE_NODE_TYPES } from '@services/AWSInstanceTypes'
import { AwsDecorator } from '@services/AwsDecorator'

export default class ElastiCache extends ServiceWithCPUUtilization {
  serviceName = 'elasticache'
  readonly aws: AwsDecorator

  constructor() {
    super()
    this.aws = new AwsDecorator()
  }

  async getUsage(startDate: Date, endDate: Date): Promise<ComputeUsage[]> {
    const cloudWatchParams = {
      StartTime: startDate,
      EndTime: endDate,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/ElastiCache} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const metricDataResponses = await this.aws.getMetricDataResponses(cloudWatchParams)

    const costExplorerParams = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'USAGE_TYPE_GROUP', Values: ['ElastiCache: Running Hours'] } },
          { Dimensions: { Key: 'REGION', Values: [config.region] } },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['UsageQuantity'],
    }
    const costAndUsageResponses = await this.aws.getCostAndUsageResponses(costExplorerParams)

    return getComputeUsage(metricDataResponses, costAndUsageResponses, CACHE_NODE_TYPES)
  }
}
