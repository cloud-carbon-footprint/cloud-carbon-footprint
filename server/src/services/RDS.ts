import ServiceWithCPUUtilization from "@domain/ServiceWithCPUUtilization";
import ComputeUsage from "@domain/ComputeUsage";
import UsageData from "@domain/UsageData";

export class RDSService extends ServiceWithCPUUtilization {
    serviceName: string;

    constructor() {
        super()
    }

    getUsage(start: Date, end: Date): Promise<ComputeUsage[]>{
        return undefined;
    }

}
