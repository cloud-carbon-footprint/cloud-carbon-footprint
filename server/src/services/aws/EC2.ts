/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import ComputeUsage, { buildComputeUsages, extractRawComputeUsages, RawComputeUsage } from '@domain/ComputeUsage'
import Cost from '@domain/Cost'
import { GetCostAndUsageRequest } from 'aws-sdk/clients/costexplorer'
import { getCostFromCostExplorer } from '@services/aws/CostMapper'
import { MetricDataResult } from 'aws-sdk/clients/cloudwatch'
import { ServiceWrapper } from '@services/aws/ServiceWrapper'

export default class EC2 extends ServiceWithCPUUtilization {
  serviceName = 'ec2'

  constructor(private serviceWrapper: ServiceWrapper) {
    super()
  }

  async getUsage(start: Date, end: Date): Promise<ComputeUsage[]> {

    const response = await this.serviceWrapper.getQueryByInterval(30, this.runQuery.bind(this), start, end)
    const flattenedResp = response.reduce((acc, data) => [...data, ...acc], [])
    return flattenedResp
  }

  private async runQuery(start: Date, end: Date): Promise<ComputeUsage[]>{
    const params = {
      StartTime: start,
      EndTime: end,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
        {
          Id: 'vCPUs',
          Expression:
            'SEARCH(\'{AWS/Usage,Resource,Type,Service,Class } Resource="vCPU" MetricName="ResourceCount"\', \'Average\', 3600)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const responses = await this.serviceWrapper.getMetricDataResponses(params)

    const metricDataResults: MetricDataResult[] = responses.flatMap((response) => response.MetricDataResults)
    
    const rawComputeUsages: RawComputeUsage[] = metricDataResults.flatMap(extractRawComputeUsages)
    return buildComputeUsages(rawComputeUsages, 'AWS')
  }

  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {

    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          {
            Dimensions: {
              Key: 'REGION',
              Values: [region],
            },
          },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['EC2: Running Hours'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
      Metrics: ['AmortizedCost'],
    }
    return getCostFromCostExplorer(params, this.serviceWrapper)
  }
}
