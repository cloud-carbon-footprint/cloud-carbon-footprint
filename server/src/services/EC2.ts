import AWS from 'aws-sdk'
import ServiceWithCPUUtilization from '@domain/ServiceWithCPUUtilization'
import ComputeUsage from '@domain/ComputeUsage'

export default class EC2 extends ServiceWithCPUUtilization {
  serviceName = 'ec2'
  readonly cloudWatch: AWS.CloudWatch

  constructor() {
    super()
    this.cloudWatch = new AWS.CloudWatch()
  }

  async getUsage(startDate: Date, endDate: Date): Promise<ComputeUsage[]> {
    const params = {
      StartTime: startDate,
      EndTime: endDate,
      MetricDataQueries: [
        {
          Id: 'cpuUtilizationWithEmptyValues',
          Expression: "SEARCH('{AWS/EC2,InstanceId} MetricName=\"CPUUtilization\"', 'Average', 3600)",
          ReturnData: false,
        },
        {
          Id: 'cpuUtilization',
          Expression: 'REMOVE_EMPTY(cpuUtilizationWithEmptyValues)',
        },
        {
          Id: 'vCPUs',
          Expression:
            'SEARCH(\'{AWS/Usage,Resource,Type,Service,Class } Resource="vCPU" MetricName="ResourceCount"\', \'Average\', 3600)',
        },
      ],
      ScanBy: 'TimestampAscending',
    }

    const response = await this.cloudWatch.getMetricData(params).promise()

    interface RawComputeUsage {
      cpuUtilization?: number[]
      vCPUCount?: number
      timestamp?: Date
      cpuUtilizationAvg?: number
    }
    const result: { [key: string]: RawComputeUsage } = {}

    // Aggregate CPU Utilization
    const metricDataResults = response.MetricDataResults
    const cpuUtilizationData = metricDataResults.filter((a) => a.Id === 'cpuUtilization')
    cpuUtilizationData.forEach((instanceCPUUtilization) => {
      instanceCPUUtilization.Timestamps.forEach((timestamp, i) => {
        const key = new Date(timestamp).toISOString()
        if (!result[key]) result[key] = { cpuUtilization: [instanceCPUUtilization.Values[i]], timestamp: timestamp }
        result[key].cpuUtilization.push()
      })
    })

    // Add vCPU to result object
    const vCPUData = metricDataResults.filter((a) => a.Id === 'vCPUs')
    vCPUData[0]?.Timestamps.forEach((timestamp, i) => {
      const key = new Date(timestamp).toISOString()
      if (!result[key]) result[key] = { cpuUtilization: [], timestamp: timestamp }
      result[key].vCPUCount = vCPUData[0].Values[i]
    })

    // Apply estimation formula
    Object.values(result).forEach((a) => {
      a.cpuUtilizationAvg =
        a.cpuUtilization.reduce((acc, a) => {
          return acc + a
        }, 0) / a.cpuUtilization.length

      if (isNaN(a.cpuUtilizationAvg)) {
        a.cpuUtilizationAvg = 0.0
      }
    })

    // Aggregate by date
    const estimationsByDay: { [timestamp: string]: ComputeUsage } = {}
    Object.values(result).forEach((a) => {
      const timestamp = new Date(a.timestamp)
      const date = timestamp.toISOString().substr(0, 10)

      if (!estimationsByDay[date]) {
        estimationsByDay[date] = {
          timestamp: new Date(date),
          cpuUtilizationAverage: a.cpuUtilizationAvg,
          numberOfvCpus: a.vCPUCount,
        }
      } else {
        estimationsByDay[date].cpuUtilizationAverage += a.cpuUtilizationAvg
        estimationsByDay[date].numberOfvCpus += a.vCPUCount
      }
    })

    return Object.values(estimationsByDay)
  }
}
