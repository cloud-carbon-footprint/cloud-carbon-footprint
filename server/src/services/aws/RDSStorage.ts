import { GetCostAndUsageRequest } from 'aws-sdk/clients/costexplorer'
import ICloudService from '@domain/ICloudService'
import {
  DiskType,
  getEstimatesFromCostExplorer,
  getUsageFromCostExplorer,
  VolumeUsage,
} from '@services/aws/StorageUsageMapper'
import FootprintEstimate from '@domain/FootprintEstimate'
import Cost from '@domain/Cost'
import { getCostFromCostExplorer } from '@services/aws/CostMapper'
import { ServiceWrapper } from '@services/aws/ServiceWrapper'

export default class RDSStorage implements ICloudService {
  serviceName = 'rds-storage'

  constructor(private readonly serviceWrapper: ServiceWrapper) {}

  async getEstimates(start: Date, end: Date, region: string): Promise<FootprintEstimate[]> {
    const usage: VolumeUsage[] = await this.getUsage(start, end, region)
    return getEstimatesFromCostExplorer(start, end, region, usage)
  }

  async getUsage(startDate: Date, endDate: Date, region: string): Promise<VolumeUsage[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: startDate.toISOString().substr(0, 10),
        End: endDate.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'REGION', Values: [region] } },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['RDS: Storage'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['UsageQuantity'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return await getUsageFromCostExplorer(params, this.getDiskType, this.serviceWrapper)
  }

  private getDiskType = (awsGroupKey: string) => {
    if (awsGroupKey.endsWith('GP2-Storage') || awsGroupKey.endsWith('PIOPS-Storage')) return DiskType.SSD
    if (awsGroupKey.endsWith('StorageUsage')) return DiskType.HDD
    console.warn('Unexpected Cost explorer Dimension Name: ' + awsGroupKey)
  }

  async getCosts(start: Date, end: Date, region: string): Promise<Cost[]> {
    const params: GetCostAndUsageRequest = {
      TimePeriod: {
        Start: start.toISOString().substr(0, 10),
        End: end.toISOString().substr(0, 10),
      },
      Filter: {
        And: [
          { Dimensions: { Key: 'REGION', Values: [region] } },
          {
            Dimensions: {
              Key: 'USAGE_TYPE_GROUP',
              Values: ['RDS: Storage'],
            },
          },
        ],
      },
      Granularity: 'DAILY',
      Metrics: ['AmortizedCost'],
      GroupBy: [
        {
          Key: 'USAGE_TYPE',
          Type: 'DIMENSION',
        },
      ],
    }

    return getCostFromCostExplorer(params, this.serviceWrapper)
  }
}
