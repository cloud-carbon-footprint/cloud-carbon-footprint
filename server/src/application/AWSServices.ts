import CloudService from '@domain/CloudService'
import EBS from '@services/EBS'
import S3 from '@services/S3'
import EC2 from '@services/EC2'
import ElastiCache from '@services/ElastiCache'
import RDSComputeService from '@services/RDSCompute'

export default function AWSServices(): CloudService[] {
  return [new EBS(), new S3(), new EC2(), new ElastiCache(), new RDSComputeService()]
}
