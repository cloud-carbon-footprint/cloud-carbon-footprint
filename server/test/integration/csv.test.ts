import { mocked } from 'ts-jest/utils'
import AWSServices from '@application/AWSServices'
import EBS from '@services/EBS'
import S3 from '@services/S3'
import EC2 from '@services/EC2'
import ElastiCache from '@services/ElastiCache'
import RDS from '@services/RDS'
import RDSComputeService from '@services/RDSCompute'
import RDSStorage from '@services/RDSStorage'
import path from 'path'
import fs from 'fs'
import { mockAwsCloudWatchGetMetricData, mockAwsCostExplorerGetCostAndUsage } from '../fixtures/awsMockFunctions'
import cli from '@view/cli'
import AWSMock from 'aws-sdk-mock'
import AWS from 'aws-sdk'

jest.mock('@application/AWSServices')

//disable cache
jest.mock('@application/Cache')

beforeAll(() => {
  AWSMock.setSDKInstance(AWS)
})

afterEach(() => {
  AWSMock.restore()
})

describe('csv test', () => {
  const start = '2020-07-01'
  const end = '2020-07-07'
  const servicesRegistered = mocked(AWSServices, true)
  const rawRequest = ['executable', 'file', '--startDate', start, '--endDate', end, '--region', 'us-east-1']

  servicesRegistered.mockReturnValue([
    new EBS(),
    new S3(),
    new EC2(),
    new ElastiCache(),
    new RDS(new RDSComputeService(), new RDSStorage()),
  ])

  let outputFilePath: string

  beforeEach(() => {
    jest.spyOn(Date, 'now').mockImplementation(() => 1596660091000)
    outputFilePath = path.join(process.cwd(), 'results-2020-08-05-20:41:31.csv')
  })

  afterEach(() => {
    try {
      fs.unlinkSync(outputFilePath)
    } catch (err) {
      console.error(err)
    }
    jest.restoreAllMocks()
  })

  test('formats table into csv file', async () => {
    mockAwsCloudWatchGetMetricData()
    mockAwsCostExplorerGetCostAndUsage()

    await cli([...rawRequest, '--format', 'csv', '--groupBy', 'dayAndService'])

    expect(fs.existsSync(outputFilePath)).toBe(true)
    expect(fs.readFileSync(outputFilePath).toString()).toMatchSnapshot()
  })
})
