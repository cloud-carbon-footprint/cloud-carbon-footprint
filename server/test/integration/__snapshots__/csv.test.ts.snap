// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csv test formats table into csv file 1`] = `
"Date (UTC),EBS Watt Hours,EBS Kg CO2e Emissions,S3 Watt Hours,S3 Kg CO2e Emissions,EC2 Watt Hours,EC2 Kg CO2e Emissions,ElastiCache Watt Hours,ElastiCache Kg CO2e Emissions,RDS Watt Hours,RDS Kg CO2e Emissions,Lambda Watt Hours,Lambda Kg CO2e Emissions,ComputeEngine Watt Hours,ComputeEngine Kg CO2e Emissions,SUM Watt Hours,SUM Kg CO2e Emissions
2020-07-01,1.07,0.000361,0.05,0.000017,19.55,0.006585,2.28,0.000767,3.97,0.001336,0.00,0.000000,0.00,0.000000,26.91,0.009066
2020-07-02,2.14,0.000722,0.00,0.000000,0.00,0.000000,9.38,0.003162,226.29,0.076243,0.00,0.000000,0.00,0.000000,237.81,0.080126
Total,3.21,0.001083,0.05,0.000017,19.55,0.006585,11.66,0.003928,230.25,0.077579,0.00,0.000000,0.00,0.000000,264.72,0.089192
"
`;
